ui/src/router.tsx

```diff
@@ -0,0 +1,94 @@
+import Button from "@mui/material/Button";
+import Edit from "edit";
+import { PathError } from "hooks";
+import Overview from "overview";
+import { ErrorBoundary } from "react-error-boundary";
+import {
+  Route,
+  Switch,
+  useHistory,
+  useParams,
+  useRouteMatch,
+} from "react-router-dom";
+import { UnderlaySelect } from "underlaySelect";
+import { Datasets } from "./datasets";
+
+const pages = {
+  "/": UnderlaySelect,
+  "/:underlayName": Datasets,
+  "/:underlayName/cohorts/:cohortId": Overview,
+  "/:underlayName/cohorts/:cohortId/edit/:groupId/:criteriaId": Edit,
+};
+
+export function AppRouter() {
+  return (
+    <ErrorBoundary FallbackComponent={NotFound}>
+      <Switch>
+        {Object.entries(pages).map(([page, component]) => (
+          <Route exact key={page} path={page} component={component} />
```

Comment:
I'm surprised typescript likes this. You didn't need to apply stronger explicit typing to `pages`?

Reply:
All of the components take no props so they have the same type. It make sense since they're all top level pages that fetch their data from the the URL.